File: samples/tab-stage-view/nodejs/server/bot/botActivityHandler.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { AgentApplication, MemoryStorage, MessageFactory } = require('@microsoft/agents-hosting');
const { TeamsAgentExtension } = require('@microsoft/agents-hosting-extensions-teams');
const adaptiveCards = require('../models/adaptiveCard');
const adaptiveCardUnfurling = require('../models/adaptiveCardUnfurling');

/**
 * Creates and configures the Agent Application with Teams extension
 * @returns {AgentApplication} The configured agent application
 */
function createAgentApp() {
    const agent = new AgentApplication({
        storage: new MemoryStorage()
    });

    // Register Teams extension for Teams-specific features like link unfurling
    const teamsExt = new TeamsAgentExtension(agent);

    agent.registerExtension(teamsExt, (tae) => {
        // Handle app based link query (link unfurling)
        tae.messageExtension.onQueryLink(async (context, state, link) => {
            console.log('Link unfurling triggered for URL:', link);
            const card = adaptiveCardUnfurling.adaptiveCardForTabStageView();
            return {
                attachmentLayout: 'list',
                type: 'result',
                attachments: [{
                    contentType: 'application/vnd.microsoft.card.adaptive',
                    content: card,
                    preview: {
                        contentType: 'application/vnd.microsoft.card.thumbnail',
                        content: {
                            title: 'Tab Stage View',
                            text: 'Click to open in stage view',
                            images: []
                        }
                    }
                }]
            };
        });
    });

    // Handle invoke activities for link unfurling (fallback handler)
    agent.onActivity('invoke', async (context, state) => {
        const activity = context.activity;
        console.log('Invoke activity received:', activity.name);
        
        if (activity.name === 'composeExtension/queryLink') {
            console.log('Link unfurling request:', activity.value?.url);
            const card = adaptiveCardUnfurling.adaptiveCardForTabStageView();
            
            const response = {
                composeExtension: {
                    attachmentLayout: 'list',
                    type: 'result',
                    attachments: [{
                        contentType: 'application/vnd.microsoft.card.adaptive',
                        content: card,
                        preview: {
                            contentType: 'application/vnd.microsoft.card.adaptive',
                            content: card
                        }
                    }]
                }
            };
            
            await context.sendActivity({
                type: 'invokeResponse',
                value: {
                    status: 200,
                    body: response
                }
            });
            return;
        }
    });

    // Handle new members added to conversation
    agent.onConversationUpdate('membersAdded', async (context) => {
        const welcomeText = "Hello and welcome!, Please type any bot command to see the stage view feature";
        await context.sendActivity(welcomeText);
    });

    // Handle incoming messages
    agent.onActivity('message', async (context, state) => {
        // Base url without protocol to be used in OpenUrl encoded deeplink
        const baseUrl = process.env.BaseUrl || process.env.BOT_ENDPOINT || '';
        const card = adaptiveCards.adaptiveCardForTabStageView(baseUrl);
        
        const message = MessageFactory.attachment({
            contentType: 'application/vnd.microsoft.card.adaptive',
            content: card
        });
        
        await context.sendActivity(message);
    });

    return agent;
}

module.exports = { createAgentApp };

File: samples/tab-stage-view/python/bots/bot_activity_handler.py
import os
from botbuilder.core import TurnContext, MessageFactory
from botbuilder.schema import Activity, ActivityTypes
from botbuilder.core.teams import TeamsActivityHandler
from botbuilder.schema.teams import AppBasedLinkQuery, MessagingExtensionResult, MessagingExtensionAttachment, MessagingExtensionResponse
from botbuilder.schema import Attachment

from models import adaptive_card, adaptive_card_unfurling


class BotActivityHandler(TeamsActivityHandler):
    def __init__(self):
        super().__init__()

    async def on_message_activity(self, turn_context: TurnContext):
        base_url = os.getenv("BaseUrl", "")
        base_url = base_url.split("://")[-1] if "://" in base_url else base_url

        card = adaptive_card.adaptive_card_for_tab_stage_view(base_url)
        attachment = Attachment(
            content_type="application/vnd.microsoft.card.adaptive",
            content=card
        )

        await turn_context.send_activity(Activity(
            type=ActivityTypes.message,
            attachments=[attachment]
        ))

    async def on_members_added_activity(self, members_added, turn_context: TurnContext):
        welcome_text = "Hello and welcome! Please type any bot command to see the stage view feature"
        await turn_context.send_activity(MessageFactory.text(welcome_text))

    async def on_teams_app_based_link_query(self, turn_context: TurnContext, query: AppBasedLinkQuery):
        card = adaptive_card_unfurling.adaptive_card_for_tab_stage_view()
        attachment = MessagingExtensionAttachment(
            content_type="application/vnd.microsoft.card.adaptive",
            content=card
        )

        result = MessagingExtensionResult(
            type="result",
            attachment_layout="list",
            attachments=[attachment]
        )

        return MessagingExtensionResponse(compose_extension=result)
